### PR TITLE
Make control_point take references to LocPhases, not LocPhases themselves

### PR DESCRIPTION
**Pull request by ltratt**
Make control_point take references to LocPhases, not LocPhases themselves.

We initially inherited the RPython idea that the way to express a control_point is to take in a location in the end user's program and then see if the hot threshold for that point has been exceeded or not. The good thing about this is that it's a simple user API. However, underneath, we use a hashtable: every time you call the control_point, you have to pay the costs of calculating a hash and performing a hash lookup. These are fairly cheap operations, but they're not free, and in the context of a tight interpreter loop they're likely to be noticeable. To add insult to injury, in our multi-threading context, a thread-safe hashmap adds additional costs. For users who want to squeeze out every last bit of performance, this is not great.

Fortunately, there seems to me to be an obvious solution. Instead of taking in a user program's location and internally storing a LocPhase, we ask users to give us a reference to the LocPhase corresponding to the current location. At the very worst: for a bytecode interpreter, a user can simply allocate a Vec with one entry per bytecode; for an AST interpreter a LocPhase per node.

At its simplest, the API looks as follows:

```rust
  let lp = LocPhase::new();
  mt.control_point(&lp);
```

For users who care more about memory usage, it isn't difficult to imagine some simple changes to a language's compiler. For example, if we have a loop like:

```rust
  for x in 0..10 {
     ...
  }
```

it would traditionally end up being transformed into bytecode such as:

```
      SET x, 0
  L0: ...
      ADD x, 1
      CMP x, 10
      JLT L0
```

We know that the interpreter only needs to call control_point at L0 (since nowhere else in that snippet can be the start of a loop). We could, for example, add a new bytecode whose sole job is to call control_point e.g.:

```
      SET x, 0
  L0: START_LOOP
      ...
      ADD x, 1
      CMP x, 10
      JLT L0
```

Since this is fairly easy for a compiler to work this out, it should also be fairly easy to store a LocPhase corresponding to that START_LOOP bytecode.

This API also allows us to free ourselves from the idea that there has to be a single lock for all LocPhases. What this commit first does is to turn LocPhases from an enum into a packed integer with a tag (the top 2 bits) and value (the other 30 bits). We can then uses atomic integers to reduce the read/write costs to as low as is practical. On an amd64 machine, the "common" case (which is that there is a compiled trace available) is exactly equivalent to a normal load (i.e. there are no CAS-esque costs and no possibility of threads starving each other out). [We do have to deal with the possibility that, while we're incrementing hot counts, multiple threads can be operating on the same LocPhase: fortunately a simple spin-lock-esque deals with this nicely. If you crank up the DEFAULT_HOT_THRESHOLD, you can see races happen, and dealt with correctly.]

Let me put some rough performance numbers on this. There are two Cargo benchmarks in this PR. On bencher8 they run thus:

```
  test metatracer::tests::bench_multi_threaded_control_point  ... bench:     148,637 ns/iter (+/- 8,489)
  test metatracer::tests::bench_single_threaded_control_point ... bench:      55,782 ns/iter (+/- 514)
```

I've ported those two benchmarks to the code-before-this-commit (always using a single location "0") and they run thus:

```
  test metatracer::tests::bench_multi_threaded_control_point  ... bench:  68,497,983 ns/iter (+/- 1,715,450)
  test metatracer::tests::bench_single_threaded_control_point ... bench:   4,100,744 ns/iter (+/- 3,549)
```

That means that the multi_threaded benchmark has sped up ~450x and the single_threaded benchmark has sped up ~75x.

Working out what part of that speed-up is the lack of locking and what's down to dictionaries is tricky. As a rough (rough!) experiment, I took the code-before-this-commit and removed the dictionary lookups (i.e. doing the mutex locking, but not the dict lookup) and got. The new multi_threaded then becomes only ~270x faster and the new single_threaded only ~30x faster. That suggests that the speed-up is roughly split 50/50 between removing locking and removing dictionary lookups.

Summary: we allow interpreter authors who care about performance to avoid hashtable costs; and we make performance about as good on a multi-threaded system as I know how to do.

### CODE REVIEW

**Comment by ltratt** [comment]
As an incidental detail, this PR resolves https://github.com/softdevteam/yorickrt/issues/4 -- I needed the functionality to make tests.

**Comment by vext01** [comment]
Renaming the repo has confused travis. I think you might need to close and re-open your PR.

**Comment by ltratt** [comment]
I've done a force push, which looks like it's done a right thing.

**Comment by vext01** [comment]
I've not really looked at your bit hacking or CAS logic yet, but I think understand where you are aiming with this change. I must admit it took quite some time to understand what was going on here.

Let's see if I got it right. In my own words.

Instead of storing a hash table of location states, we ask the user to take storage of this stuff into their own hands. The most obvious storage strategy seems to be for the user to store a vector of locations. The user then passes references to these into our control point and we mutate it and potentially change phase depending upon what we see inside.

The advantage seems to be that there is not hash lookup any more, which seems to speed us up a bit (but as you say it's hard to pick that apart from contention costs). The downside is that every possible location from which a loop may start must always be in memory.

If that's right, I think it's quite clever once you get your head around it.

Of course I have questions:

 * The API is kind of "inside-out" to what users will expect. Do you think they will accept it?
 * How many loop headers would a typical medium-sized  program have to store in our locations vec? And how much memory would that consume?
 * For a language like PHP, which has per-function program counters, would we have one vec per function? In a way, the problem of specifying reds and greens disappears. Did you notice?

Another day I need to look closer at the nitty-gritty details.

**Comment by ltratt** [comment]
>The downside is that every possible location from which a loop may start must always be in memory.

I'm not sure what this means. If you mean "must always have an associated LocPhase" -- yes. But there again, in RPython, you'd have a hash entry (which probably takes more memory than a LocPhase), so I don't think it's any different.


>    The API is kind of "inside-out" to what users will expect. Do you think they will accept it?

I think people will be OK with the new API. If I'm wrong, I could provide an RPython style dict thingy, but it'll take more memory and be slower than having a `Vec` of `LocPhase`s. [I was going to add this in as a helper before I came to the conclusion that a `Vec` of `LocPhase`s isn't too onerous.]

>    How many loop headers would a typical medium-sized program have to store in our locations vec?

I don't know what this means.

>    For a language like PHP, which has per-function program counters, would we have one vec per function?

A function entry point would just be another `LocPhase`.

>In a way, the problem of specifying reds and greens disappears. Did you notice?

Greens (or whichever one the program location is!) disappear entirely (at least, I don't need them yet: I can't totally rule out needing them at some point in the future). Reds don't disappear as such, though I think we might not need to make the user explicitly pass them in (which is a bit neater, I admit).

**Comment by ltratt** [comment]
Oops, I forgot to be clear about this:

>Instead of storing a hash table of location states, we ask the user to take storage of this stuff into their own hands. The most obvious storage strategy seems to be for the user to store a vector of locations. The user then passes references to these into our control point and we mutate it and potentially change phase depending upon what we see inside.

Yes, that's completely correct!

**Comment by vext01** [comment]
>  If you mean "must always have an associated LocPhase" -- yes. But there again, in RPython, you'd have a hash entry (which probably takes more memory than a LocPhase), so I don't think it's any different.

Yep, that's what I meant.

> 

>> How many loop headers would a typical medium-sized program have to store in our locations vec?
>
> I don't know what this means.

I was trying to get an idea how much memory the vectors of `LocPhase`s would get. But judging by your last answer, I think "not bigger than RPython" is your answer? In which case, it's probably OK.

> Greens (or whichever one the program location is!) disappear entirely (at least, I don't need them yet

I think as long as we can always quickly derive an integer index into our vec, we don't need any explicit greens.

Note that the user could have multiple vecs of `LocPhase` if they so wished. At least I think your API allows it.

I think I'm happy to give this a shot, but please wait for me to read up on the atomics etc. I've never used them.

**Comment by ltratt** [comment]
>I was trying to get an idea how much memory the vectors of LocPhases would get. But judging by your last answer, I think "not bigger than RPython" is your answer?

I don't know if RPython does any clever bit packing (or similar) but, assuming not, then I think "not bigger than RPython" is probably true.

>I think as long as we can always quickly derive an integer index into our vec, we don't need any explicit greens.

Well, maybe there are other things we'd need greens for, but I don't *think* so. At least, I can't think of anything yet.

>Note that the user could have multiple vecs of LocPhase if they so wished. At least I think your API allows it.

Yes, they can do anything they like provided they pass in a `&LocPhase`.

**Comment by vext01** [comment]
Last thought before zzz. Should we `s/LocPhase/Loc/g`. I know I proposed that name, but now I'm starting to dislike it.

**Comment by ltratt** [comment]
I don't feel strongly either way.

**Comment by vext01** [comment]
Let's leave it for now, but have a think about it.

**Comment by ltratt** [comment]
On reflection, I agree that `Loc` is a better name since it's more generic (the user doesn't need to know that we're storing phase information in it). I'll push a commit to that effect shortly.

**Comment by ltratt** [comment]
There's a commit (OK, I chose "Location" over "Loc", but that's not hugely significant).

File: src/metatracer.rs [hunk 1]
```diff
@@ -42,14 +42,17 @@ const PHASE_COUNTING: u32 = 0b00; // The value specifies the current hot count.
 const PHASE_TRACING : u32 = 0b01;
 const PHASE_COMPILED: u32 = 0b10;
 
-/// A LocPhase stores data that the `MetaTracer` needs internally to determine if and when to trace
-/// a given location, and whether a compiled trace already exists for that location.
-pub struct LocPhase {
+/// A `Location` identifies a control point position in the end-user's program (the `MetaTracer`
```
Comment (vext01) [inline src/metatracer.rs]: Shall we mention that the identity of a `Location` matters? That passing two equivalent `Locations` will not do what the user expects?

Obvious to us, I know.

**Comment by vext01** [comment]
One boring comment on that last commit.

File: src/metatracer.rs (hunk 1 above)
Comment (ltratt) [reply src/metatracer.rs]: How about "A `Location`uniquely identifies"?

**Comment by ltratt** [comment]
OK, I've clarified that comment.

What do you want to do with the main bulk of the PR?

File: src/metatracer.rs (hunk 1 above)
Comment (vext01) [reply src/metatracer.rs]: Is that strictly true? I'd say that the address of a `Location` uniquely identifies it, but even that is horrible to offer up to a user as an explanation.

I guess I'm trying to guard against people thinking that they could materialise `Locations` on the fly without storing them. I could totally see myself try something silly like that.

**Comment by vext01** [comment]
I need to check your concurrency stuff. This might be a task for the airport.

File: src/metatracer.rs (hunk 1 above)
Comment (ltratt) [reply src/metatracer.rs]: Yes, it's true. [The fact that we only let people give us references to the location is kind-of secondary.] But you're right, it could be misread. Let me try again.

Comment (ltratt) [reply src/metatracer.rs]: [Indeed, we don't care if you change the address of a Location, provided you give us the same contents.]

Comment (vext01) [reply src/metatracer.rs]: good point.

**Comment by ltratt** [comment]
OK, see if the new commit clarifies things.

**Comment by vext01** [comment]
Better. Two small suggestions:

 * "and that `Location` must always be used to identify that control point." -> "and that *same* `Location` must always be used to identify that control point."
 * If we mention that the profiling and trace info is inside would this help users see why they can't make new ones all the time?

**Comment by ltratt** [comment]
OK, try that.

File: src/metatracer.rs [hunk 2]
```diff
@@ -30,57 +30,89 @@
...
+impl MetaTracer {
     pub fn new() -> Self {
         Self {
-            counts: Mutex::new(HashMap::new())
+            hot_threshold: AtomicU32::new(DEFAULT_HOT_THRESHOLD)
         }
     }
 
+    /// Update the hot threshold for this meta-tracer instance to `threshold`.
+    pub fn set_hot_threshold(&mut self, threshold: HotThreshold) {
```
Comment (vext01) [inline src/metatracer.rs]: Do we expect the user to be able to change the hot threshold after initialising the JIT?

If not, then why don't we kill this function, have the hot threshold as a parameter to `new()`, and then the hot count can be a regular `u32`, thus simplifying the code in the control point (no need for the atomic load/store dance).

Comment (ltratt) [reply src/metatracer.rs]: Good point. Yes, that would simplify things.

Comment (ltratt) [reply src/metatracer.rs]: [Well, I think the default constructor should still take no parameters; but there's no harm in having a second constructor.]

File: src/metatracer.rs [hunk 3]
```diff
@@ -30,57 +30,89 @@
...
+                PHASE_COUNTING => {
+                    let count = (lp & !PHASE_TAG) >> PHASE_TAG;
+                    let new_pack;
+                    if count >= self.hot_threshold.load(Ordering::Relaxed) {
+                        new_pack = PHASE_TRACING;
+                    } else {
+                        new_pack = PHASE_COUNTING | ((count + 1) << PHASE_TAG);
+                    }
+                    if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
+                        break;
```
Comment (vext01) [inline src/metatracer.rs:107]: OK, so if I understand this correctly, this is a lock-free update to the pack. The atomics will:
 * inform the compiler to respect the ordering.
 * inform the hardware to respect the ordering by emitting fences into the compiled code.

The breaks from the loop took me a while to understand. We exit our loop if the old value was the same old value we observed before we tried to update the pack. If this was not the case, that suggests that another thread updated the pack in the meantime, and that we should try again.

Have I understood this correctly?

Did you base this section of code on an example you found somewhere on the internet? If so, could you share the link?

Comment (ltratt) [reply src/metatracer.rs:107]: Yes, that's a correct understanding of my intention. No, it's not based on other code.

**Comment by vext01** [comment]
It looks OK to me, but I don't claim to be an expert on atomics.

This is the first time I've seen CAS and atomic orderings in the wild.

I wonder if there's someone we can ask to review this section? Just for the
added confidence that we got it right.

**Comment by ltratt** [comment]
I don't know anyone who it would be reasonable to ask to review this. And even if we did, we've got more concurrency stuff coming up, so I think we have to expect to take on the burden ourselves... We'll learn, maybe after a mistake or two, but we'll learn :)

File: src/metatracer.rs [hunk 4]
```diff
@@ -64,19 +64,20 @@ impl Location {
...
-        Self {
-            hot_threshold: AtomicU32::new(DEFAULT_HOT_THRESHOLD)
-        }
+        Self::new_with_hot_threshold(DEFAULT_HOT_THRESHOLD)
     }
 
-    /// Update the hot threshold for this meta-tracer instance to `threshold`.
-    pub fn set_hot_threshold(&mut self, threshold: HotThreshold) {
-        self.hot_threshold.store(threshold, Ordering::Relaxed);
+    /// Create a new `MetaTracer` with a specific hot threshold `hot_threshold`.
```
Comment (vext01) [inline src/metatracer.rs]: I'd probably kill the second backticked `hot_threshold` since it's obvious, but we are well into the realms of nitpicking now :)

**Comment by vext01** [comment]
I think I'm good with this! I'll leave it up to you if you want to address my nitpick, otherwise, let's squash and merge?

**Comment by ltratt** [comment]
If you're happy with that latest commit then I agree, it's squashing time.

**Comment by vext01** [comment]
go for it

**Comment by ltratt** [comment]
Squashed.

**Comment by vext01** [comment]
Didn't you rename `LocPhase`? The commit still has a reference to this terminology.

**Comment by ltratt** [comment]
It was called `LocPhase` before, so at least one of the references is correct (since it refers to the old thing). I tried to rename things properly, but please check to see if I did it correctly!

**Review by vext01 — APPROVED** [review]
